### PR TITLE
add windows native aligned malloc + unit test case

### PR DESCRIPTION
**Pull request by diablodale**
Implements https://github.com/opencv/opencv/issues/19147 by adding native Win32 memory alignment to the opencv core memory allocator and test case. Test data collected prior to PR suggests allocation might be 2% slower. Core maintainer prefers this code be considered for merge. Refer to issue for much detail on accuracy and performance data.

### Pull Request Readiness Checklist

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [X] The PR is proposed to proper branch
- [X] There is reference to original bug report and related work
- [X] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [X] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/core/test/test_allocation.cpp
Comment (alalek) [inline modules/core/test/test_allocation.cpp:11]: Please move this test into `test_misc.cpp` and keep make it disabled (add `DISABLED_` prefix).

**Review by alalek — COMMENTED** [review]
Well done! Thank you for contribution :+1:

**Comment by alalek** [comment]
This patch should go into [3.4 branch](https://github.com/opencv/opencv/wiki/Branches) first.
We will merge changes from 3.4 into master regularly (weekly/bi-weekly).

Please:
- change "base" branch of this PR: master => 3.4 (use "Edit" button near PR title)
- rebase your commits from master onto 3.4 branch. For example:
    `git rebase -i --onto upstream/3.4 upstream/master`
    (check list of your commits, save and quit (Esc + "wq" + Enter)
    where `upstream` is configured by following [this GitHub guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and fetched (`git fetch upstream`).
- push rebased commits into source branch of your fork (with `--force` option)

Note: no needs to re-open PR, apply changes "inplace".

File: modules/core/test/test_allocation.cpp
Comment (diablodale) [reply modules/core/test/test_allocation.cpp:14]: I separately tested for fragmentation. Please see the issue for detailed analysis.

File: modules/core/test/test_allocation.cpp
Comment (diablodale) [reply modules/core/test/test_allocation.cpp:11]: I can move it, no problem. However, I believe it is a good idea to keep the test enabled. The time cost is less than a millisecond and the value is testing the core memory allocator for all of opencv. That's value..yes?

Comment (alalek) [reply modules/core/test/test_allocation.cpp:11]: > The time cost is less than a millisecond

OK, Agreed.

---

> memory allocator for all of opencv

From code coverage perspective it is already tested everywhere.

**Comment by diablodale** [comment]
> This patch should go into [3.4 branch](https://github.com/opencv/opencv/wiki/Branches) first.
> We will merge changes from 3.4 into master regularly (weekly/bi-weekly).
> 
> Please:
> 
> * change "base" branch of this PR: master => 3.4 (use "Edit" button near PR title)
> * rebase your commits from master onto 3.4 branch. For example:
>   `git rebase -i --onto upstream/3.4 upstream/master`
>   (check list of your commits, save and quit (Esc + "wq" + Enter)
>   where `upstream` is configured by following [this GitHub guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and fetched (`git fetch upstream`).
> * push rebased commits into source branch of your fork (with `--force` option)
> 
> Note: no needs to re-open PR, apply changes "inplace".

Won't be able to do that. Several issues:

1. This PR is 4.5 and greater only. It will not work on 3.x. It requires the earlier PR to the 4.x branch that was merged that fixes the startup behavior in this translation unit. It absolutely 100% will not work in 3.x
2. All the testing, performance, and fragmentation analysis I have done was using the 4.x commit that is listed in the issue. Those test results and analysis can not be used to make decisions about 3.x.
3. OpenCV 3.x is bugfix only. That is everywhere in official documentation. This PR is a new feature and therefore for the 4.x branch.

Usually, I will *not* be submitting PRs for the legacy 3.x branch. All the things I do (code, test, analysis, etc.) will be on the current 4.x branch. I do not have the bandwidth to do all the work with a broken 22 year old standard in a branch declared as bugfix-only. I understand you are a maintainer of this project. And I'm a contributor. We have to have a middle-ground where contributions work for us both. 🤝

**Comment by alalek** [comment]
OK, I will backport this separately after merge.

**Comment by diablodale** [comment]
You wrote 
> OK, I will backport this separately after merge.

I discourage and caution you not to do this. Why? It is unsafe and you risk Windows crashes.

The native Windows aligned APIs used in this PR are **unsafe** until the thread-safe startup bugs of the core opencv allocator are fixed. One collection of these bugs are https://github.com/opencv/opencv/issues/19004. It is a complex bug to fix correctly if you don't use C++11. You would need to create critical sections and/or mutex to protect in needed areas, and put cascaded try/catch handlers throughout the codepath. And all that has to be done very carefully. It will likely reduce performance.

Only after those multiple unsafe (thread and exception unsafe) allocator bugs are fixed using C++98, can this PR's code then be applied. If you don't fix those unsafe bugs, then you risk a hard crash of Windows applications. Windows will not allow a mix of deallocators `free` and `_aligned_free`.  Linux is different, it only uses `free`.

We have a saying in English, "Let sleeping dogs lie." 🐶 It means: if there isn't a current problem...and a change is challenging...it is wise to not make a change. I believe that applies with opencv 3.x.

The existing opencv 3.x codebase *on windows* does not have the startup bug 19004. This is because that 3.x codebase uses `ifdef` on Windows to remove all the buggy code, uses fixed branches, and uses the handmade opencv alignment function. All of this works. And doesn't have any startup issues. I believe the complexity and challenge to fix 19004 in C++98 just to get this PR is not wise.

**Review by alalek — APPROVED** [review]

**Pull request merged**
